File: app/src/main/java/github/umer0586/sensorserver/broadcastreceiver/BroadcastMessageReceiver.java
package github.umer0586.sensorserver.broadcastreceiver;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.util.Log;

import androidx.annotation.NonNull;

import github.umer0586.sensorserver.service.SensorService;


public class BroadcastMessageReceiver extends BroadcastReceiver {

    private static final String TAG = BroadcastMessageReceiver.class.getSimpleName();

    private MessageListener messageListener;

    private Context context;

    private boolean isRegistered = false;


    public BroadcastMessageReceiver(@NonNull Context context)
    {
        this.context = context;
    }

    @Override
    public void onReceive(Context context, Intent intent)
    {
        Log.d(TAG, "onReceive() called with: context = [" + context + "], intent = [" + intent + "]");

        // Dispatch received broadcast intent via MessageListener Interface
        if(messageListener != null)
            messageListener.onMessage(intent);

    }

    public void setMessageListener(MessageListener messageListener)
    {
        this.messageListener = messageListener;
    }

    public void registerEvents()
    {
        Log.d(TAG, "registerEvents() called");

        IntentFilter intentFilter = new IntentFilter();

        intentFilter.addAction(SensorService.ACTION_STOP_SERVER);


        try
        {
            if(!isRegistered)
                this.context.registerReceiver(this, intentFilter);

            isRegistered = true;
        }catch(IllegalArgumentException e)
        {
            isRegistered = false;
            e.printStackTrace();
        }
    }

    public void unregisterEvents()
    {
        Log.d(TAG, "unregister() called");
        try
        {
            if (isRegistered)
                this.context.unregisterReceiver(this);

            isRegistered = false;
        }catch(IllegalArgumentException e)
        {
            isRegistered = false;
            e.printStackTrace();
        }
    }


    public interface MessageListener {
        public void onMessage(Intent intent);
    }
}

File: app/src/main/java/github/umer0586/sensorserver/broadcastreceiver/BroadcastMessageReceiver.kt
package github.umer0586.sensorserver.broadcastreceiver

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.util.Log
import androidx.core.content.ContextCompat

class BroadcastMessageReceiver(private val context: Context) : BroadcastReceiver()
{
    private var messageReceiveCallBack : ((Intent) -> Unit)? = null
    private var isRegistered = false

    companion object
    {
        private val TAG: String = BroadcastMessageReceiver::class.java.simpleName
    }

    fun setOnMessageReceived(callBack : ((Intent) -> Unit)?)
    {
        messageReceiveCallBack = callBack
    }

    override fun onReceive(context: Context, intent: Intent)
    {
        Log.d(TAG, "onReceive() intent = [$intent]")
        messageReceiveCallBack?.invoke(intent)

    }

    fun registerEvents(intentFilter: IntentFilter)
    {
        Log.d(TAG, "registerEvents() called")

        try
        {
            if (!isRegistered)
                ContextCompat.registerReceiver(context,this, intentFilter,ContextCompat.RECEIVER_NOT_EXPORTED)


            isRegistered = true
        }
        catch (e: IllegalArgumentException)
        {
            isRegistered = false
            e.printStackTrace()
        }
    }

    fun unregisterEvents()
    {
        Log.d(TAG, "unregister() called")

        try
        {
            if (isRegistered)
                context.unregisterReceiver(this)

            isRegistered = false
        }
        catch (e: IllegalArgumentException)
        {
            isRegistered = false
            e.printStackTrace()
        }
    }


}